src/nghttp.cc

```diff
@@ -414,6 +482,14 @@ int submit_request(HttpClient *client, const Headers &headers, Request *req) {
 
   req->req_nva = std::move(build_headers);
 
+  if (expect_continue) {
+    auto timer = std::make_shared<ContinueTimer>(client->loop, req);
+    timer->start();
+
+    req->continue_timer = timer;
+    client->continue_timers.push_back(timer);
```

Comment:
> There is one more race condition when timer is fired, before HEADERS frame is actually sent. [...] To workaround this, instead of starting timer just after nghttp2_submit_headers(), do it in nghttp2_on_frame_send_callback().

Ah, I bet that's the same race I was forcing with the low timeout. I will give that (and the rest of your suggestions) a try; thanks!